### PR TITLE
dev-python/*: version bumps

### PR DESCRIPTION
**Pull request by arthurzam**
A Pull Request with various version bumps for python packages. This is continuation of mailing with 
@mgorny .

1. All were tested on `python2_7`, `python3_7`, `pypy`, `pypy3` and passed successfully.
2. All were bumped to `EAPI=7`
3. I have tried to correct `BDEPEND`. I know that `dev-python/setuptools` should be a `BDEPEND`, but not sure about others like `dev-python/pytest`, `dev-python/nose`. What should be done about test dependencies?
4. When `IUSE="test"` I added to new `ebuilds`: `RESTRICT="!test? ( test )"`.
5. One point of concern is `KEYWORDS=amd64-linux`. `dev-lang/python` doesn't support this keyword, but a lot of `dev-python/*` packages set it.

### CODE REVIEW

**Comment by mgorny** [comment]
Without reviewing yet, few notes:
1. Whenever possible, you may try the new `distutils_enable_tests` API.
2. Test-deps go into BDEPEND.
3. Do all of those packages have working tests? Any of them need test support added?
4. Don't worry about Prefix keywords. Prefix team takes care of updating them as necessary. They may look weird, they combine their own overlay with ::gentoo.

**Comment by arthurzam** [comment]
> Without reviewing yet, few notes:
> 
>     1. Whenever possible, you may try the new `distutils_enable_tests` API.

When changed to use this API, have added a message to commit.

>     2. Test-deps go into BDEPEND.
Done
>     3. Do all of those packages have working tests? Any of them need test support added?
Yes. For some of `dev-python/zope-*` packages I have added tests, and is documented in the commit message.
>     4. Don't worry about Prefix keywords. Prefix team takes care of updating them as necessary. They may look weird, they combine their own overlay with ::gentoo.

**Comment by arthurzam** [comment]
Remove patches for `dev-python/{attrs,pluggy,hypothesis}`, as @mgorny had already bumped the versions.

File: dev-python/cffi/cffi-1.13.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0/${PV}"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~m68k ~mips ~ppc ~ppc64 ~s390 ~sh ~sparc ~x86 ~x64-cygwin ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
+IUSE="doc test"
+RESTRICT="!test? ( test )"
+
+RDEPEND="
+	virtual/libffi:=
```
Comment (mgorny) [inline dev-python/cffi/cffi-1.13.2.ebuild]: This is a DEPEND, not BDEPEND.

File: dev-python/cffi/cffi-1.13.2.ebuild [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	dev-python/pycparser[${PYTHON_USEDEP}]"
+BDEPEND="${RDEPEND}
+	virtual/pkgconfig
+	doc? ( dev-python/sphinx )
```
Comment (mgorny) [inline dev-python/cffi/cffi-1.13.2.ebuild]: I've screwed up here as it uses autodoc. Could you rewrite it to use the usually any_dep/check_deps voodoo? Or just rebase on mgorny:distutils-sphinx branch and use the new fancy API.

File: dev-python/cffi/cffi-1.13.2.ebuild [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	test? ( dev-python/pytest[${PYTHON_USEDEP}] )"
+
+# Avoid race on _configtest.c (distutils/command/config.py:_gen_temp_sourcefile)
```
Comment (mgorny) [inline dev-python/cffi/cffi-1.13.2.ebuild]: I suppose this might no longer be necessary.

File: dev-python/cffi/cffi-1.13.2.ebuild [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+python_compile_all() {
+	if use doc ; then
+		emake -C doc html
+		HTML_DOCS=( doc/build/html/. )
+	fi
+}
+
+python_test() {
+	"${PYTHON}" -c "import _cffi_backend as backend" || die
```
Comment (mgorny) [inline dev-python/cffi/cffi-1.13.2.ebuild]: EPYTHON here.

File: dev-python/atomicwrites/atomicwrites-1.3.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+DESCRIPTION="Atomic file writes"
+HOMEPAGE="https://github.com/untitaker/python-atomicwrites"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="MIT"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~hppa ~ia64 ~mips ~ppc ~ppc64 ~x86 ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
+IUSE="doc"
+
+# TODO: remove impl limiters when sphinx is py38
```
Comment (mgorny) [inline dev-python/atomicwrites/atomicwrites-1.3.0.ebuild]: This is stale, please remove.

File: dev-python/lazy-object-proxy/lazy-object-proxy-1.4.3.ebuild [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	# No need to benchmark
+	sed \
+		-e '/benchmark/s:test_:_&:g' \
+		-e '/pytest.mark.benchmark/d' \
+		-i tests/test_lazy_object_proxy.py || die
+	distutils-r1_python_prepare_all
+}
+
+python_test() {
+	py.test -v -v --ignore=src || die "Fails for ${EPYTHON}"
```
Comment (mgorny) [inline dev-python/lazy-object-proxy/lazy-object-proxy-1.4.3.ebuild]: Is this ignore really necessary? Maybe we could reuse `distutils_enable_tests` if not.

File: dev-python/pycryptodome/pycryptodome-3.9.4.ebuild [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+DESCRIPTION="A self-contained cryptographic library for Python"
+HOMEPAGE="https://www.pycryptodome.org https://github.com/Legrandin/pycryptodome https://pypi.org/project/pycryptodome/"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+
+LICENSE="BSD-2 Unlicense"
+SLOT="0"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~mips ~ppc ~ppc64 ~s390 ~sparc ~x86 ~ppc-aix ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~m68k-mint ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
+
+RDEPEND="dev-libs/gmp:0
+	virtual/python-cffi[${PYTHON_USEDEP}]
```
Comment (mgorny) [inline dev-python/pycryptodome/pycryptodome-3.9.4.ebuild]: This is BDEPEND, not DEPEND.

File: dev-python/zope-interface/zope-interface-4.7.1.ebuild [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+SRC_URI="mirror://pypi/${MY_PN:0:1}/${MY_PN}/${MY_P}.tar.gz"
+
+LICENSE="ZPL"
+SLOT="0"
+KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~m68k ~mips ~ppc ~ppc64 ~s390 ~sh ~sparc ~x86 ~amd64-linux ~x86-linux ~ppc-macos ~x64-macos ~x86-macos ~sparc-solaris ~sparc64-solaris ~x64-solaris ~x86-solaris"
+IUSE="test"
+RESTRICT="!test? ( test )"
+
+RDEPEND="dev-python/namespace-zope[${PYTHON_USEDEP}]"
+DEPEND="${RDEPEND}"
```
Comment (mgorny) [inline dev-python/zope-interface/zope-interface-4.7.1.ebuild]: Seems unnecessary.

File: dev-python/zope-interface/zope-interface-4.7.1.ebuild [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+python_compile() {
+	if ! python_is_python3; then
+		local CFLAGS="${CFLAGS}" CXXFLAGS="${CXXFLAGS}"
+		append-flags -fno-strict-aliasing
+	fi
+
+	distutils-r1_python_compile
+}
+
+distutils_enable_tests setup.py
```
Comment (mgorny) [inline dev-python/zope-interface/zope-interface-4.7.1.ebuild]: Move it above phases.

File: dev-python/zope-exceptions/zope-exceptions-4.3.ebuild [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+LICENSE="ZPL"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+RESTRICT="!test? ( test )"
+
+RDEPEND="dev-python/namespace-zope[${PYTHON_USEDEP}]
+	dev-python/zope-interface[${PYTHON_USEDEP}]"
+DEPEND="${RDEPEND}"
```
Comment (mgorny) [inline dev-python/zope-exceptions/zope-exceptions-4.3.ebuild]: Shouldn't be necessary.

File: dev-python/zope-i18nmessageid/zope-i18nmessageid-5.0.0.ebuild [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+DESCRIPTION="Zope i18nmessageid Architecture"
+HOMEPAGE="https://github.com/zopefoundation/zope.i18nmessageid http://docs.zope.org/zope.i18nmessageid/"
+SRC_URI="mirror://pypi/${MY_PN:0:1}/${MY_PN}/${MY_P}.tar.gz"
+
+LICENSE="ZPL"
+SLOT="0"
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc64 ~x86"
+
+RDEPEND="dev-python/six[${PYTHON_USEDEP}]"
+DEPEND="${RDEPEND}"
```
Comment (mgorny) [inline dev-python/zope-i18nmessageid/zope-i18nmessageid-5.0.0.ebuild]: Seems unnecessary.

File: dev-python/zope-schema/zope-schema-4.9.3.ebuild [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+LICENSE="ZPL"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+IUSE="test"
+RESTRICT="!test? ( test )"
+
+RDEPEND="dev-python/zope-event[${PYTHON_USEDEP}]
+	dev-python/zope-interface[${PYTHON_USEDEP}]"
+BDEPEND="dev-python/setuptools[${PYTHON_USEDEP}]
+	test? ( ${RDEPEND}
```
Comment (mgorny) [inline dev-python/zope-schema/zope-schema-4.9.3.ebuild]: d_e_t already takes care of that.

File: dev-python/zope-deprecation/zope-deprecation-4.4.0.ebuild [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+DESCRIPTION="Zope Deprecation Infrastructure"
+HOMEPAGE="https://github.com/zopefoundation/zope.deprecation"
+SRC_URI="mirror://pypi/${MY_PN::1}/${MY_PN}/${MY_P}.tar.gz"
+
+LICENSE="ZPL"
+SLOT="0"
+KEYWORDS="~amd64 ~x86"
+
+RDEPEND="dev-python/namespace-zope[${PYTHON_USEDEP}]"
+DEPEND="${RDEPEND}"
```
Comment (mgorny) [inline dev-python/zope-deprecation/zope-deprecation-4.4.0.ebuild]: Unnecessary.

File: dev-python/lazy-object-proxy/lazy-object-proxy-1.4.3.ebuild (hunk 6 above)
Comment (arthurzam) [reply dev-python/lazy-object-proxy/lazy-object-proxy-1.4.3.ebuild]: This ignore is indeed necessary, without it it tries to `import cext` which fails on `python3`. On theirs testing CI, they also use `--ignore=src`.

**Comment by arthurzam** [comment]
Thanks for review - I hope I will know all in future so it will be easier.

1. Wasn't sure about `distutils_enable_tests setup.py` setting `IUSE` and `RESTRICT`. But after rechecking if `${RDEPEND}` isn't empty it sets it - nicely done :+1: 
2. Added `python3_8` to `pydiff` (needed for `autopep8`) - of course have checked with tests.

**Comment by mgorny** [comment]
I'm going to push distutils_enable_sphinx shortly.

**Comment by arthurzam** [comment]
Rebased over new `distutils_enable_sphinx`.
Needed to change only for `dev-python/{atomicwrites,cffi}`.
For every package I touch in this PR, fixed `QA: MissingTestRestrict` (watched through [here](https://qa-reports.gentoo.org/output/gentoo-ci/output.verbose.html)).

**Comment by mgorny** [comment]
zope and imagesize done too. zope.schema doesn't run any tests.

**Comment by arthurzam** [comment]
@mgorny , Thank you - Closing now.

1. Is it a good idea to add docs support for all python packages that have `sphinx` docs?
2. How should I add `python3_8` support for `dev-python/namespace-*` packages?

**Comment by mgorny** [comment]
I didn't merge pyflakes & pycryptodome. Those two need fixes/rebase.

Namespace jus can be added when needed.

**Pull request closed without merging**